libafl/src/stages/calibrate.rs

```diff
@@ -0,0 +1,213 @@
+    phantom: PhantomData<(C, E, EM, I, O, OT, S, T, Z)>,
+}
+
+// The number of times we run the program in the calibration stage
+const CAL_STAGE_MAX: usize = 8;
+
+impl<C, E, EM, I, O, OT, S, T, Z> Stage<E, EM, S, Z>
+    for CalibrationStage<C, E, EM, I, O, OT, S, T, Z>
+where
+    T: Integer + Default + Copy + 'static + serde::Serialize + serde::de::DeserializeOwned,
+    C: Corpus<I>,
+    E: Executor<EM, I, S, Z> + HasObservers<I, OT, S>,
+    I: Input,
+    O: MapObserver<T>,
+    OT: ObserversTuple<I, S>,
+    S: HasCorpus<C, I> + HasMetadata,
+    Z: Evaluator<E, EM, I, S>,
+{
+    #[inline]
+    #[allow(clippy::let_and_return)]
+    fn perform(
+        &mut self,
+        fuzzer: &mut Z,
+        executor: &mut E,
+        state: &mut S,
+        manager: &mut EM,
+        corpus_idx: usize,
+    ) -> Result<(), Error> {
+        let iter = self.stage_max;
+        let handicap = state
+            .metadata()
+            .get::<PowerScheduleStats>()
+            .ok_or_else(|| Error::KeyNotFound("PowerScheduleStats not found".to_string()))?
+            .queue_cycles;
+
+        // Timer start
+        let start = current_time();
+
+        for _i in 0..iter {
+            let input = state
+                .corpus()
+                .get(corpus_idx)?
+                .borrow_mut()
+                .load_input()?
+                .clone();
+            let (_, _) = fuzzer.evaluate_input(state, executor, manager, input)?;
+        }
+        // Timer end
+        let end = current_time();
+
+        let map = executor
+            .observers()
+            .match_name::<O>(&self.map_observer_name)
+            .ok_or_else(|| Error::KeyNotFound("MapObserver not found".to_string()))?;
+
+        let bitmap_size = map.count_bytes();
+
+        let calstat = state
+            .metadata_mut()
+            .get_mut::<PowerScheduleStats>()
+            .ok_or_else(|| Error::KeyNotFound("PowerScheduleStats not found".to_string()))?;
+
+        calstat.set_exec_time(calstat.exec_time() + (end - start));
+        calstat.set_cycles(calstat.cycles() + (iter as u64));
+        calstat.set_bitmap_size(calstat.bitmap_size() + bitmap_size);
+        calstat.set_bitmap_entries(calstat.bitmap_entries() + 1);
+
+        // println!("calstat: {:#?}", calstat);
+        let mut testcase = state.corpus().get(corpus_idx)?.borrow_mut();
+
+        testcase.set_exec_time((end - start) / (iter as u32));
+        // println!("time: {:#?}", testcase.exec_time());
+        let data = testcase
+            .metadata_mut()
+            .get_mut::<PowerScheduleTestData>()
+            .ok_or_else(|| Error::KeyNotFound("PowerScheduleTestData not found".to_string()))?;
+
+        data.set_bitmap_size(bitmap_size);
+        data.set_handicap(handicap);
+        data.set_fuzz_level(data.fuzz_level() + 1);
+        // println!("data: {:#?}", data);
+
+        Ok(())
+    }
+}
+
+#[derive(Serialize, Deserialize, Clone, Debug)]
+pub struct PowerScheduleStats {
```

Comment:
Stats is used for the stats reported to the user, maybe change this to PowerScheduleMetadata